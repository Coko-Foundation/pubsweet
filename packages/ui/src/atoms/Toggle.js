/* eslint-disable sort-keys */
/* eslint-disable react/prop-types */
import React, { useState, useEffect } from 'react'
import PropTypes from 'prop-types'
import styled from 'styled-components'
import { override, th } from '@pubsweet/ui-toolkit'

const Root = styled.div`
  display: flex;
  padding: 10px;
  ${override('ui.Toggle')};
`

const ToggleUI = styled.div`
  /* color: #171717; */
  width: 37.71px;
  height: 6.08px;
  border-radius: ${th('gridUnit')};
  background: ${th('colorBackgroundHue')};
  border: 1px solid ${th('colorBackgroundHue')};
  position: relative;
  cursor: pointer;
  margin: ${th('gridUnit')};
`

const ToggleButton = styled.div`
  position: absolute;
  top: -${th('gridUnit')};
  right: ${props => (props.checked ? '-2px' : '50%')};
  width: 21px;
  height: 21px;
  background: ${props =>
    props.checked ? props.theme.colorPrimary : props.theme.colorSecondary};
  border-radius: 50%;
  transition: right 0.3s;
`

const Label = styled.div`
  padding-top: 10px;
  font-family: ${th('fontInterface')};
  font-size: ${th('fontSizeBase')};
  line-height: ${th('gridUnit')};
`

const Input = styled.input`
  visibility: hidden;
  width: 1px;
  position: absolute;
`

const Toggle = ({ label, labelChecked, name, disabled, value, onClick }) => {
  const [ischecked, setChecked] = useState(value || false)

  useEffect(() => {
    setChecked(value)
  }, [value])

  return (
    <Root>
      <Input name={name} type="text" value={ischecked ? 'true' : 'false'} />

      <ToggleUI
        checked={ischecked}
        onClick={() => {
          if (!disabled) {
            onClick(!ischecked)
            setChecked(!ischecked)
          }
        }}
      >
        <ToggleButton checked={ischecked} />
      </ToggleUI>
      {labelChecked && ischecked ? (
        <>
          <Label>{labelChecked}</Label>
        </>
      ) : (
        <>
          <Label>{label}</Label>
        </>
      )}
    </Root>
  )
}

Toggle.propTypes = {
  value: PropTypes.bool,
  disabled: PropTypes.bool,
  label: PropTypes.string,
  labelChecked: PropTypes.string,
  onClick: PropTypes.func,
  name: PropTypes.string,
}

Toggle.defaultProps = {
  value: false,
  disabled: false,
  label: '',
  labelChecked: '',
  name: '',
  onClick: () => {},
}

export default Toggle
