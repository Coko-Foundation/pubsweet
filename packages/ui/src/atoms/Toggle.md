A component that can acts as switch input.  
Often useful in forms, but not limited to them.

Basic Toggle

```js
<Toggle />
```

Toggle with label

```js
<Toggle label="Admin" />
```

Toggle with LabelChecked

```js
<Toggle label="Not Checked" labelChecked="Is Checked" />
```
