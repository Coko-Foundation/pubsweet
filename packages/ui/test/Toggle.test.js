import React from 'react'
import renderer from 'react-test-renderer'
import 'jest-styled-components'

import Toggle from '../src/atoms/Toggle'
import TestThemeProvider from './setup/theme'

const props = {
  value: false,
  disabled: false,
  label: 'Not Admin',
  labelChecked: 'Is admin',
}

describe('Toggle', () => {
  test('is rendered correctly', () => {
    const tree = renderer
      .create(
        <TestThemeProvider>
          <Toggle {...props} />
        </TestThemeProvider>,
      )
      .toJSON()
    expect(tree).toMatchSnapshot()
  })
})
