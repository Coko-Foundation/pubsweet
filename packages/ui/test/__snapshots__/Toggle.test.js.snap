// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Toggle is rendered correctly 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  padding: 8px;
}

.c0:hover span {
  color: black;
}

.c2 {
  width: calc(5 * 8px);
  height: 8px;
  border-radius: 8px;
  background: grey;
  border: 1px solid grey;
  cursor: pointer;
  margin: 8px;
}

.c2:before {
  content: ' ';
  display: inline-block;
  position: relative;
  width: calc(3 * 8px);
  height: calc(3 * 8px);
  border-radius: 50%;
  top: -8px;
  left: 0px;
  background: grey;
  -webkit-transition: left 0.3s;
  transition: left 0.3s;
}

.c1 {
  position: absolute;
  opacity: 0;
  z-index: -1;
  margin-right: 8px;
}

.c3 {
  position: absolute;
  padding-left: calc(4 * 8px);
  margin-top: -8px;
}

<label
  className="c0"
>
  <input
    checked={false}
    className="c1"
    disabled={false}
    name=""
    onChange={[Function]}
    type="checkbox"
    value={false}
  />
  <span
    checked={false}
    className="c2"
  >
    <span
      className="c3"
    >
      Not Admin
    </span>
  </span>
</label>
`;
